tests/integration/test_upgrades.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Ranger charm upgrades integration tests."""
+
+import asyncio
+import logging
+
+import pytest
+import pytest_asyncio
+import requests
+import yaml
+from helpers import APP_NAME, METADATA, POSTGRES_NAME, SECURE_PWD, get_unit_url
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
+@pytest_asyncio.fixture(name="deploy", scope="module")
+async def deploy(ops_test: OpsTest):
+    """Deploy the app."""
+    asyncio.gather(
+        ops_test.model.deploy(POSTGRES_NAME, channel="14", trust=True),
+    )
+
+    ranger_config = {"ranger-admin-password": SECURE_PWD}
+    await ops_test.model.deploy(
+        APP_NAME, channel="stable", config=ranger_config
```

Comment:
This charm would first be pushed to the `edge` channel after merging with main. Does it make sense for us to test that the new changes in `edge` do not break the `stable` release, or rather test that the latest changes do not break whatever is in the `edge` channel currently? Just thinking out loud.

Reply:
I think you're right, it makes more sense to check that this does not break the `edge` release, since edge release deployments would be updated with `juju refresh` to this version we are testing.